### DISCUSSION TITLE
Writing datarray/dataset to netcdf results in bad values

### DISCUSSION THREAD

**Discussion by etsmith14** [pydata/xarray#7025] [Q&A] [answered]
I am trying to save a regional subset of a netcdf file as netcdf file. I am first opening some data with dimensions of time, latitude, and longitude and then slicing that data by latitude and longitude to produce a smaller subset of the data. I save that smaller subset with the to_netcdf command. But when I go to open the new netcdf, the timeseries definitely wrong (see figures). The figure named 'correct' is what the temperature timeseries looks like when plotting directly from the original dataset. The figure named 'wrong' is what the temperature timeseries looks like when plotting from the newly saved netcdf (hopefully both figures attached properly).  This happens when I select just a single point and save the data as a netcdf and it also happens when I save as a zarr file. However, when I load a single netcdf with open_dataset (instead of open_mfdataset) and save it as a new netcdf, everything is correct. So the issue seems to be coming from open_mfdataset. I've also noticed that not all grid points are incorrect, only some grid points have this issue. This doesn't happen when I convert to a series then save as a CSV, just happens when saving as a netcdf or zarr.  

```python
import xarray as xr
import matplotlib.pyplot as plt

lats = [33.35]
lons = [-112.86]

# Load data from original files
ERA5_t2m = xr.open_mfdataset(r'E:\ERA5\Temperature\T2m_*' + '.nc', parallel=True).sel(latitude = slice(37.25,31), longitude = slice(-115, -109))
# plot original data (looks good)
plt.plot((((ERA5_t2m.t2m.sel(latitude = lats[0], longitude = lons[0], method='nearest')- 273.15) * 9/5) + 32))
```

![Correct](https://user-images.githubusercontent.com/35741277/189739924-3e5d3838-e6af-4453-bb33-87fc470e75c1.png)

```python
# save regional subset as new netcdf
ERA5_t2m.t2m.to_netcdf(r"E:\Arizona_test\Arizona_Temperature.nc")
# open new netcdf regional subset
ERA5_t2m_AZ = xr.open_dataset(r'E:\Arizona_test\Arizona_Temperature.nc')
# plot same point from new netcdf (looks bad)
plt.plot((((ERA5_t2m_AZ.t2m.sel(latitude = lats[0], longitude = lons[0], method='nearest')- 273.15) * 9/5) + 32))
```

![wrong](https://user-images.githubusercontent.com/35741277/189739923-77e5af32-8452-481e-a6e2-c2e6571b65ce.png)

**Comment by rabernat** [comment]
That definitely seems weird. I have a couple of hunches about what might be going wrong, but they're hard to confirm with more detail.

Can you start by sharing more detailed metadata about the two datasets?

```python
print(ERA5_t2m)
print(ERA5_t2m_AZ)
```

and also

```python
print(ERA5_t2m.t2m.attrs)
print(ERA5_t2m_AZ.t2m.attrs)
print(ERA5_t2m.t2m.encoding)
print(ERA5_t2m_AZ.t2m.encoding)
```

**Reply by etsmith14** [reply]
Thanks for responding to my issue so quickly! Here is the print outs:

**print(ERA5_t2m)**
`<xarray.Dataset>
Dimensions:    (longitude: 25, latitude: 26, time: 631152)
Coordinates:
  * longitude  (longitude) float32 -115.0 -114.8 -114.5 ... -109.5 -109.2 -109.0
  * latitude   (latitude) float32 37.25 37.0 36.75 36.5 ... 31.5 31.25 31.0
  * time       (time) datetime64[ns] 1950-01-01 ... 2021-12-31T23:00:00
Data variables:
    t2m        (time, latitude, longitude) float32 dask.array<chunksize=(744, 26, 25), meta=np.ndarray>
    mx2t       (time, latitude, longitude) float32 dask.array<chunksize=(744, 26, 25), meta=np.ndarray>
    mn2t       (time, latitude, longitude) float32 dask.array<chunksize=(744, 26, 25), meta=np.ndarray>
Attributes:
    Conventions:  CF-1.6
    history:      2022-05-03 17:18:58 GMT by grib_to_netcdf-2.24.3: /opt/ecmw...`


**print(ERA5_t2m_AZ)**
`<xarray.Dataset>
Dimensions:    (longitude: 25, latitude: 26, time: 631152)
Coordinates:
  * longitude  (longitude) float32 -115.0 -114.8 -114.5 ... -109.5 -109.2 -109.0
  * latitude   (latitude) float32 37.25 37.0 36.75 36.5 ... 31.5 31.25 31.0
  * time       (time) datetime64[ns] 1950-01-01 ... 2021-12-31T23:00:00
Data variables:
    t2m        (time, latitude, longitude) float32 ...`

**print(ERA5_t2m.t2m.attrs)**
`{'units': 'K', 'long_name': '2 metre temperature'}`


**print(ERA5_t2m_AZ.t2m.attrs)**
`{'units': 'K', 'long_name': '2 metre temperature'}`

**print(ERA5_t2m.t2m.encoding)**
`{'source': 'E:\\ERA5\\Temperature\\T2m_19501.nc', 'original_shape': (744, 109, 245), 'dtype': dtype('int16'), 'missing_value': -32767, '_FillValue': -32767, 'scale_factor': 0.0011997040993123216, 'add_offset': 269.40377331689564}`


**print(ERA5_t2m_AZ.t2m.encoding)**
`{'zlib': False, 'szip': False, 'zstd': False, 'bzip2': False, 'blosc': False, 'shuffle': False, 'complevel': 0, 'fletcher32': False, 'contiguous': True, 'chunksizes': None, 'source': 'E:\\Arizona_test\\Arizona_Temperature.nc', 'original_shape': (631152, 26, 25), 'dtype': dtype('int16'), 'missing_value': -32767, '_FillValue': -32767, 'scale_factor': 0.0011997040993123216, 'add_offset': 269.40377331689564}`

**Comment by rabernat** [comment, marked as answer]
Ok thanks. I don't see anything obvious yet. Let's try one more.

```python
print(ERA5_t2m.t2m.sel(latitude = lats[0], longitude = lons[0], method='nearest')
print(ERA5_t2m_AZ.t2m.sel(latitude = lats[0], longitude = lons[0], method='nearest')
```

**Reply by etsmith14** [reply]
**print(ERA5_t2m.t2m.sel(latitude = lats[0], longitude = lons[0], method='nearest')**
<xarray.DataArray 't2m' (time: 631152)>
dask.array<getitem, shape=(631152,), dtype=float32, chunksize=(744,), chunktype=numpy.ndarray>
Coordinates:
    longitude  float32 -112.8
    latitude   float32 33.25
  * time       (time) datetime64[ns] 1950-01-01 ... 2021-12-31T23:00:00
Attributes:
    units:      K
    long_name:  2 metre temperature




**print(ERA5_t2m_AZ.t2m.sel(latitude = lats[0], longitude = lons[0], method='nearest')**
<xarray.DataArray 't2m' (time: 631152)>
[631152 values with dtype=float32]
Coordinates:
    longitude  float32 -112.8
    latitude   float32 33.25
  * time       (time) datetime64[ns] 1950-01-01 ... 2021-12-31T23:00:00
Attributes:
    units:      K
    long_name:  2 metre temperature

**Reply by etsmith14** [reply]
I'll also add, I've tried rolling xarray back several versions and I've tried installing xarray and all it's dependencies in a fresh venv with the same issue.

**Reply by rabernat** [reply]
This is pretty bizarre.

Two more questions:
- Can you reproduce the problem with just one file, i.e. `xr.open_dataset(r'E:\ERA5\Temperature\T2m_000.nc')`. And if so, can you share the file?
- Have you examined systematically the difference between the two? Like, is there a constant offset, etc? That might help narrow down the problem.

**Reply by etsmith14** [reply]
No, the problem doesn't occur with just one file, just when I do open_mfdataset. However, if I loop through every file and open it using open_dataset, then append the data to a list and concatenate by 'time', then the issue does happen.

**Reply by etsmith14** [reply]
Not sure I follow the second question. Constant offset in the coordinates? To be honest, from the print encoding above, I am not sure what  the 'add_offset': 269.40377331689564 means. 

Also, I've saved regional netcdf subsets before and never noticed this issue.

**Reply by rabernat** [reply]
I'm suggesting you try to examine `ERA5_t2m.t2m - ERA5_t2m_AZ.t2m` to understand the nature of the difference. (Ignore the attrs and encoding for now.) Just have a look. Make a few lat / lon pcolormesh plots. Look a a few timeseries. Is there a pattern to the difference? Just running through some idea I would follow personally to get to the bottom.

**Reply by rabernat** [reply]
My other suggestion would be try try completely removing the encoding before saving, i.e. 

```python
ERA5_t2m.t2m.encoding = {}
ERA5_t2m.to_netcdf(...)
```

**Reply by etsmith14** [reply]
Gotcha. I'll give that try and report back!

**Reply by etsmith14** [reply]
I just tried ERA5_t2m.t2m.encoding = {} before writing to netcdf and it worked! I also tried ERA5_t2m.encoding = {} and it did not work (the dataset contains the 3 variables, mx2t, mn2t, and t2m). I assume this must be applied to the individual variables? 

I'll do some more thorough checks, but so far it looks like the problem is solved! Thanks for all your help!

**Reply by etsmith14** [reply]
Update - I plotted every grid point for min, max, and mean temperature for the entire region and no issues. There is also no issues when saving to zarr anymore. Everything looks good!

**Reply by rabernat** [reply]
Ok, I'm glad we found a solution for you. But it's still a bug!

For a little more context, have a look at what the docs say about [writing encoded data](https://docs.xarray.dev/en/stable/user-guide/io.html#writing-encoded-data).

Basically, what is happening here is that:
- The original netcdf files store encoded data (scale_factor, add_offset, etc.)
- When Xarray reads the file, it _decodes_ the data variables according to this encoding and sticks the encoding information in the `.encoding` dictionary
- When Xarray writes back the file, it _re-encodes_ the data according to whatever it finds in `.encoding`. It does this so that you can "round trip" data from a netcdf file, i.e. `xr.open_dataset(file1.nc).to_netcdf(file2.nc)` should be as identical as possible.
- By deleting `.encoding`, you are telling xarray to skip the re-encoding step and just use its defaults 

Something subtle is going on around encoding and concatenation that I don't quite understand. But we should get to the bottom of it. I wish we could develop a [minimal reproducible example](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) for this bug that does not involve hundreds on netCDF files.

**Reply by etsmith14** [reply]
Thanks for the additional context. Perhaps I can create a small subset and try to reproduce the problem then pass everything along. Is there a best practice for sharing data on Github? I could send a box drive link with the files or just drag and drop a zipped folder here.

**Reply by rabernat** [reply]
Ideally the MRE would not involve any files at all but only code! Something like this:

- Create a dummy Xarray dataset from scratch with the encoding that causes the issue.
- Save to netcdf
- Re-open, and show that it is different from the origin dataset
  
However, developing this is a lot of work, which you might not have time for. If you can reproduce the issue with a few small files, then you could just attach them directly to the github issue I believe.

Thanks for taking the time to help us improve xarray.

**Reply by etsmith14** [reply]
I created a single time series of temperature for Phoenix (1950 - 2021). Phx_Temperature.nc is the netcdf with no issues, Phx_Temperature_bad.nc is the netcdf with the issues. Hopefully, this data will suffice. 

If the zip file doesn't work, I added the files to a box folder and made a publicly accessible link: https://epri.box.com/s/k0l5zfulcngim3ahluowiwlm33b5a6hd

[Test.zip](https://github.com/pydata/xarray/files/9568331/Test.zip)

**Reply by rabernat** [reply]
Thanks so much for taking the time to do that! 🙏 

I realize it was not clear what I was asking of you. In order for this to be actionable by our developers, a bit more is required.

1. Go to https://github.com/pydata/xarray/issues
2. Open an issue following the template
3. Craft a [minimal reproducible example](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) of the bug using the attached files. Provide complete code that anyone can copy and paste to reproduce the bug. The goal is to make it as simple as possible for a potential contributor to explore and diagnose the issue. This will be a modified version of the code at the top of this post.

I know this is a lot of work on your part. You may decide you don't have time, which is also fine. But this is the process that will lead most directly to the bug getting fixed. This is how open source moves forward. 🙏

**Reply by etsmith14** [reply]
Okay, will do! Thanks for clarifying!

**Reply by veenstrajelmer** [reply]
Just for completeness: https://github.com/pydata/xarray/issues/7039
